Use safe methods for get and set in free modules

cxzhong opened the pull request:
Fix #41249 Use safe methods only.
<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

dimpase reviewed (changes requested):
get() is an unsafe method, as documented. What's the point of this? Just use `__getitem__`,  which (should be, although I didn't check) is safe

cxzhong commented on src/sage/modules/free_module_element.pyx:
It use ``get_unsafe`` in ``_getitem_``

dimpase replied:
it certainly does (it's probably the only one "real" array element raw access method). I wonder why this call is not wrapped in sig_on/sig_off - is it a bug?
@nbruin - what would you think?

nbruin replied:
It looks like `get` is slower and with less features than `__getitem__`, so I suspect it can just be deprecated. In the mean time, slapping on sig_on/sig_off could be OK, but shouldn't we really just be checking bounds and raising an error? As observed, you may not get a segfault for out-of-bounds access. In that case, you'll just get a garbage value. That's even worse! The argument would usually be "but I only use this routine for in-bounds access and then it's faster". We don't seem to have the "faster" part here. So if people want faster code, they should move to `__getitem__` anyway (or cython where there's a proper "unsafe").

So it doesn't seem very important what "get" does. It might just be a nuisance. Did anyone look through its blame to see where it's coming from? Sometimes old tickets give insight or it may show that there was a problem before that has somehow been resolved in a different way.

user202729 replied:
This `get_unsafe` is guaranteed to be safe because of the `PySlice_GetIndicesEx` above. No change needed here.

tscrim reviewed (changes requested):
I don’t think we should deprecate these (as they are useful for novice users and discoverable by a documentation search as this is such a widely used area of math, in particular, students) but instead redirect to the `__getitem__` and `__setitem__` to do the safety checks.

dimpase reviewed (approved):
lgtm

tscrim reviewed (changes requested):
Again, I am staunchly opposed to deprecating these methods.

cxzhong committed:
Use safe methods for get and set in free modules

Replace direct unsafe indexing (get_unsafe/set_unsafe) with safe
alternatives (bracket access and set method) in free_module.py and
free_module_element.pyx to add bounds checking and avoid potential
segfaults. Update and consolidate doctests accordingly.

tscrim reviewed (approved):
I also agree that this change should not be contraversal.
